File: runtime-go/internal/sinks/file/file.go
package file

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/linxinhong/eventweave/runtime-go/internal/event"
)

// Sink appends events to a JSONL file.
type Sink struct {
	path       string
	outputDir  string
	file       *os.File
	count      int
	failed     int
}

// New creates a file sink constrained to outputDir.
func New(path, outputDir string) *Sink { return &Sink{path: path, outputDir: outputDir} }

// ValidatePath checks that path is contained within outputDir without side effects.
func ValidatePath(path, outputDir string) error {
	_, err := resolveWithinOutputDir(path, outputDir)
	return err
}

// resolveWithinOutputDir resolves the target path against outputDir and ensures
// the result is contained within outputDir.
func resolveWithinOutputDir(target, outputDir string) (string, error) {
	base, err := filepath.Abs(outputDir)
	if err != nil {
		return "", fmt.Errorf("resolve output directory: %w", err)
	}

	var resolved string
	if filepath.IsAbs(target) {
		resolved, err = filepath.EvalSymlinks(target)
		if err != nil {
			resolved = filepath.Clean(target)
		}
	} else {
		resolved, err = filepath.EvalSymlinks(filepath.Join(base, target))
		if err != nil {
			resolved = filepath.Clean(filepath.Join(base, target))
		}
	}

	rel, err := filepath.Rel(base, resolved)
	if err != nil || strings.HasPrefix(rel, "..") || rel == ".." {
		return "", fmt.Errorf("file sink path %q escapes output directory %q", target, outputDir)
	}
	return resolved, nil
}

// Open creates the parent directory and opens the file.
func (s *Sink) Open() error {
	resolved, err := resolveWithinOutputDir(s.path, s.outputDir)
	if err != nil {
		return err
	}
	s.path = resolved

	if err := os.MkdirAll(filepath.Dir(s.path), 0o755); err != nil {
		return err
	}
	file, err := os.OpenFile(s.path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return err
	}
	s.file = file
	return nil
}

// Write appends one event.
func (s *Sink) Write(ev event.Event) error {
	if s.file == nil {
		return fmt.Errorf("file sink is not open")
	}
	data, err := json.Marshal(ev)
	if err != nil {
		s.failed++
		return err
	}
	if _, err := s.file.Write(append(data, '\n')); err != nil {
		s.failed++
		return err
	}
	s.count++
	return nil
}

// Flush flushes the file.
func (s *Sink) Flush() error {
	if s.file == nil {
		return nil
	}
	return s.file.Sync()
}

// Close closes the file.
func (s *Sink) Close() error {
	if s.file == nil {
		return nil
	}
	return s.file.Close()
}

// Count returns emitted events.
func (s *Sink) Count() int { return s.count }

// Failed returns failed writes.
func (s *Sink) Failed() int { return s.failed }

File: eventweave/runtime/sinks/file.py
"""File sink that appends events to a JSONL file."""

from __future__ import annotations

import json
from pathlib import Path
from typing import TextIO

from eventweave.core.event import Event
from eventweave.runtime.sink import Sink


def _resolve_within_output_dir(target: Path, output_dir: Path) -> Path:
    """Resolve *target* against *output_dir* and ensure it stays within it."""
    base = output_dir.expanduser().resolve()
    resolved = target.resolve() if target.is_absolute() else (base / target).resolve()

    try:
        resolved.relative_to(base)
    except ValueError as exc:
        raise ValueError(
            f"file sink path {target} escapes output directory {output_dir}"
        ) from exc

    return resolved


class FileSink(Sink):
    """Append events to a JSONL file within a constrained output directory."""

    def __init__(
        self,
        path: str | Path,
        output_dir: str | Path,
    ) -> None:
        self.output_dir = Path(output_dir)
        self.path = _resolve_within_output_dir(Path(path), self.output_dir)
        self._count = 0
        self._file: TextIO | None = None

    def open(self) -> None:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        self._file = self.path.open("a", encoding="utf-8")

    def write(self, event: Event) -> None:
        if self._file is None:
            raise RuntimeError("FileSink is not open")
        self._file.write(json.dumps(event.model_dump(), default=str, ensure_ascii=False) + "\n")
        self._count += 1

    def flush(self) -> None:
        if self._file is not None:
            self._file.flush()

    def close(self) -> None:
        if self._file is not None:
            self._file.close()
            self._file = None

    def count(self) -> int:
        return self._count
